### ISSUE TITLE
Is my Siamese Network correct?

### ISSUE THREAD

**Issue by willduan** [tflearn/tflearn#812]
I want to write a siamese network example , and it can run, but I don't know is it correct, can anyone help me to review it?
```
from __future__ import division, print_function, absolute_import

import tflearn
from tflearn.data_utils import to_categorical, pad_sequences
from tflearn.datasets import imdb
from tflearn.layers.merge_ops import merge
from tflearn.layers.recurrent import simple_rnn
from tflearn.metrics import accuracy_op

import tensorflow as tf

# IMDB Dataset loading
train, test, _ = imdb.load_data(path='imdb.pkl', n_words=10000, valid_portion=0.1)
trainX, trainY = train
testX, testY = test
# Data preprocessing
# Sequence padding
trainX = pad_sequences(trainX, maxlen=100, value=0.)
testX = pad_sequences(testX, maxlen=100, value=0.)

def siamese_network2(input, re = False):
    with tf.variable_scope('siamese_net2', reuse = re):
    #with tf.name_scope('siamese_net2'):
        net = tflearn.embedding(input, input_dim=10000, output_dim=128)
        net = tflearn.lstm(net, 128, dropout=0.8)
        net = tflearn.fully_connected(net, 32, activation='relu')
        net = tflearn.fully_connected(net, 2, activation='relu')
        return net
#building networks
net1_input = tflearn.input_data([None, 100])
net2_input = tflearn.input_data([None, 100])

net1 = siamese_network1(net1_input)
net2 = siamese_network2(net2_input)

#define loss
distance = tf.sqrt(tf.reduce_sum(tf.square(tf.subtract(net1, net2)),1,keep_dims=True))
distance = tf.div(distance, tf.add(tf.sqrt(tf.reduce_sum(tf.square(net1),1,keep_dims=True)),tf.sqrt(tf.reduce_sum(tf.square(net2),1,keep_dims=True))))
distance = tf.reshape(distance, [-1], name="distance")

network = tflearn.regression(distance, optimizer='sgd', loss = 'contrastive_loss', metric = 'accuracy', learning_rate=0.002) #categorical_crossentropy

#training
model = tflearn.DNN(network, tensorboard_verbose = 0)
model.fit([trainX, trainX], trainY, n_epoch=5, show_metric=True, batch_size=32)
```
and the output is like this:
```
training Step: 39  | total loss: 8.31092 | time: 8.359s
| SGD | epoch: 001 | loss: 8.31092 - binary_acc: 0.5194 -- iter: 01248/22500
```
I sense it's error, can anyone help me to see that?

**Comment by willduan** [comment]
@aymericdamien

**Comment by zhao62** [comment]
I am a beginner. Did net1 and net2 share the same weights in your code? As I remember, the two sub-networks should share the same weights in siamese networks.

**Comment by ufdul** [comment]
I am trying to implement a siamese network for the first time. I don't have an experience with variable sharing. I don't know why I become this Error "Variable conv2/W does not exist, or was not created with tf.get_variable(). Did you mean to set reuse=tf.AUTO_REUSE in VarScope?"                                                  Any help will be appreciated

`from future import division, print_function, absolute_import

import tensorflow as tf

import tflearn
from tflearn.layers.core import input_data, dropout, fully_connected
from tflearn.layers.conv import conv_2d, max_pool_2d
from tflearn.layers.normalization import local_response_normalization
from tflearn.layers.estimator import regression
Data loading and preprocessing

import tflearn.datasets.mnist as mnist
X, Y, testX, testY = mnist.load_data(one_hot=True)
X = X.reshape([-1, 28, 28, 1])
testX = testX.reshape([-1, 28, 28, 1])

def tower_network(reuse= False):
network = tflearn.input_data(shape=(None,28,28,1))
network = tflearn.conv_2d(network, 32,1, activation='relu',reuse=reuse, scope='conv1')
network = tflearn.conv_2d(network, 64,1, activation='relu',reuse=reuse, scope='conv2')
network = tflearn.conv_2d(network, 128,1, activation='relu',reuse=reuse, scope='conv3')

network = tflearn.max_pool_2d(network, 2, strides=2)

network = tflearn.fully_connected(network, 512, activation='relu',reuse=reuse, scope='fc1')

network = tflearn.dropout(network, 0.5)
return network

def similarity_network( net1, net2):
num_classes = 2
network = tflearn.merge([net1,net2], mode='concat', axis=1, name='Merge') # merge net1 and net2 networks
# fully connected layers
network = tflearn.fully_connected(network, 2048, activation='relu')
network = tflearn.dropout(network, 0.5)
network = tflearn.fully_connected(network, 2048, activation='relu')
network = tflearn.dropout(network, 0.5)
# softmax layers
network = tflearn.fully_connected(network, num_classes, activation='softmax')
return network

net1 = tower_network()
net2 = tower_network(reuse=True)

#similarity network
network = similarity_network( net1, net2)
#output layer
#network = tflearn.regression(network, optimizer='sgd', loss='hinge_loss', learning_rate=0.02)
network = tflearn.regression(network, optimizer='sgd', loss='categorical_crossentropy', learning_rate=0.02)
Training

model = tflearn.DNN(network, tensorboard_verbose=0)
model.fit({'input': X}, {'target': Y}, n_epoch=20,
validation_set=({'input': testX}, {'target': testY}),
snapshot_step=100, show_metric=True, run_id='convnet_mnist')`
